### PR TITLE
Proposal for lookup_muni function

### PR DESCRIPTION
**Pull request by BAFurtado**
Rafa, per suggestion, here is a proposal for lookup_muni. Personal tests are at the end of function. A proper test file should still be developed. Best,

### CODE REVIEW

**Comment by rafapereirabr** [comment]
Thanks a bunch! We already have a shortset of tests in the `R` version at `geobr\r-package\tests\testthat\test-lookup_muni.R` . Could create the python version?

File: python-package/geobr/lookup_muni.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+""" Lookup municipality codes and names
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: Hey,

Can you follow the documentation convention from the other functions? 

Actually the function description has to be below the function statement. Take a look on the other functions :)

File: python-package/geobr/lookup_muni.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    # For the specific case in which there is just one year. For more years, change for a list result of urls
+    url = temp_meta.loc[:, 'download_path'].to_list()[0]
+
+    # Read DataFrame available at provided url
+    lookup_table_2010 = pd.read_csv(url)
+    lookup_table_2010.name_muni_format = lookup_table_2010.name_muni_format.str.lower()
+
+    # Search by inputs
+    if code_muni == 'all' or name_muni == 'all':
+        print(f"Returning results for all municipalities")
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: I avoided printing results. Python users usually don't like it.

The other functions have a `verbose` param that can be activated if the user wants to know what is happening under the hood.

File: python-package/geobr/lookup_muni.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if len(output) == 0:
+            print("Please insert a valid municipality name")
+        else:
+            print(f"Returning results for municipality {output.loc[:, 'name_muni'].to_list()[0]}")
+            return output.iloc[:, :-1]
+    elif code_muni == 'all' and name_muni == 'all':
+        print("Please insert either a municipality name or a municipality code")
+
+
+if __name__ == '__main__':
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: We don't need that in the production version.

File: python-package/geobr/utils.py [hunk 4]
```diff
@@ -240,4 +243,20 @@ def test_options(choosen, name, allowed=None, not_allowed=None):
     if not_allowed is not None:
         if choosen in not_allowed:
             raise Exception(f"Invalid value to argument '{name}'. " 
-                            f"It cannot be {' or '.join(change_type_list(allowed))}")
\ No newline at end of file
+                            f"It cannot be {' or '.join(change_type_list(allowed))}")
+
+
+def strip_accents(text):
+    """
```
Comment (JoaoCarabetta) [inline python-package/geobr/utils.py]: Please follow the documentation convention.

File: python-package/geobr/lookup_muni.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+mun_all <- lookup_muni(code_muni = "all")
+"""
+import pandas as pd
+import utils
+
+
+def lookup_muni(name_muni='', code_muni=''):
+    # Get metadata with data url addresses
+    temp_meta = utils.download_metadata()
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: Please use the function `utils.select_metadata`. It already filters by year, type and simplified. And handles errors and exceptions.

File: python-package/geobr/lookup_muni.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    temp_meta = temp_meta[(temp_meta.geo == 'lookup_muni') & (temp_meta.year == 2010)]
+
+    # For the specific case in which there is just one year. For more years, change for a list result of urls
+    url = temp_meta.loc[:, 'download_path'].to_list()[0]
+
+    # Read DataFrame available at provided url
+    lookup_table_2010 = pd.read_csv(url)
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: Maybe we can use the utils.download_metadata function here which has the same functionality, but it handles errors.

File: python-package/geobr/lookup_muni.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    lookup_table_2010.name_muni_format = lookup_table_2010.name_muni_format.str.lower()
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: please use `lookup_table_2010['name_muni_format']`, just to keep consistent with the rest of the codebase

File: python-package/geobr/lookup_muni.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Search by inputs
+    if code_muni == 'all' or name_muni == 'all':
+        print(f"Returning results for all municipalities")
+        return lookup_table_2010.iloc[:, :-1]
+    elif code_muni != '':
+        if name_muni != '':
+            print("Ignoring argument name_muni")
+        output = lookup_table_2010[lookup_table_2010.code_muni == int(code_muni)].iloc[:, :-1]
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: `lookup_table_2010['code_muni']`

File: python-package/geobr/lookup_muni.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        print(f"Returning results for municipality {output.loc[:, 'name_muni'].to_list()[0]}")
+        return output
+    elif name_muni != '':
+        # Cleaning from accents and turning into lower cases without spaces
+        name_muni = utils.strip_accents(str(name_muni).lower().strip())
+        output = lookup_table_2010[lookup_table_2010.name_muni_format == name_muni]
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: `lookup_table_2010['code_muni']`

**Comment by BAFurtado** [comment]
> Nice job!
> 
> Overall the function seems to do the job.
> 
> I am waiting for the tests and changes to approve it :)

All right. Made the corrections in the documentation format and used the suggested conventions and functions!

File: python-package/geobr/lookup_muni.py [hunk 10]
```diff
@@ -1,35 +1,44 @@
...
-Input a municipality NAME or CODE and get the names and codes of the municipality's corresponding state, meso, micro,
-intermediate, and immediate regions.
+import pandas as pd
+import utils
 
-:param name_muni -- The municipality name to be looked up
-:param code_muni -- The municipality code to be looked up
 
-:return data.frame with 13 columns identifying the geographies information of that municipality
+def lookup_muni(name_muni='', code_muni='', verbose=False):
```
Comment (JoaoCarabetta) [inline python-package/geobr/lookup_muni.py]: I have a suggestion for the conditions flow:

First, set the default to `None`
`name_muni=None` and `code_muni=None`

```python
if name_muni is None and code_muni is None:
     return <all data>
elif name_muni is not None and code_muni is not None:
     raise <Please insert either a municipality name or a municipality code>
elif name_muni is not None:
    return <filter by name muni>
elif cod_muni is not None:
   return <filter by cod_muni>
```

I think it is more pythonic than defaulting a variable to `''` and the default is to get all table.

Also, the docstring should warn the user that you should not select both `cod_muni` and `name_muni`

File: python-package/geobr/lookup_muni.py (hunk 5 above)
Comment (JoaoCarabetta) [reply python-package/geobr/lookup_muni.py]: Have you changed the function?

**Comment by BAFurtado** [comment]
@rafapereirabr @JoaoCarabetta I did the last two changes and included tests that are similar to R tests. I hope this does it. Best,

**Comment by rafapereirabr** [comment]
Thank you Bernardo! It looks good to me but I couldn't test code in action. @JoaoCarabetta , let me me if this works on your side and I'll move forward with the this PR

**Pull request merged**